### ISSUE TITLE
bazel: Investigate why sentry-cli fails to show `unwind` information for Bazel built executable

### ISSUE THREAD

**Issue by themarwhal** [magma/magma#10893]
To reproduce, it is easiest to do this in the Magma VM (might work in the devcontainer too, I haven't tested it).

First, this is the existing behavior we have.
1. Install sentry-cli: `curl -sL https://sentry.io/get-cli/ | bash`
2. build SessionD with the legacy way `cd ~/magma/lte/gateway && make build_session_manager`. This will produce a sessiond executable at `$C_BUILD/session_manager/sessiond`.
3. Run `sentry-cli difutil check $C_BUILD/session_manager/sessiond`
4. Make sure you get `unwind`, `symtab`, and `debug`.

If we try this with the Bazel built sessiond, we find that the `unwind` information is missing.
1. Install sentry-cli: `curl -sL https://sentry.io/get-cli/ | bash`
2. build SessionD with the legacy way `cd ~/magma && bazel build //lte/gateway/c/session_manager:sessiond --config=production`. (**important**: `--config=production` as the -g flag here) This will produce a sessiond executable at `~/magma/bazel-bin/lte/gateway/c/session_manager/sessiond`.
3. Run `sentry-cli difutil check ~/magma/bazel-bin/lte/gateway/c/session_manager/sessiond`
4. `unwind` is missing

I did some investigation and found that it is probably related to the linker (ld vs gold). (See more information I wrote up in Sentry's discord channel: https://discord.com/channels/621778831602221064/621802340449779720/918225454694015006
)

## Investigation and Simpler Repro Steps
Take a simple C file
```c
#include <stdio.h>
int main() {
   printf("Hello, World!\n");
   return 0;
}
```
With gcc/ld/gold
```
gcc --version
# gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
ld --version
# GNU ld (GNU Binutils for Ubuntu) 2.34
gold --version
# GNU gold (GNU Binutils for Ubuntu 2.34) 1.16
```

Running 
```bash
gcc hello.c -g -o default-linker-hello.o
```
results in the following output.
```bash
sentry-cli difutil check default-linker-hello.o
Debug Info File Check
  Type: elf library
  Contained debug identifiers:
> Debug ID: 5429ab12-bfaf-9068-0214-1c2245aa4c3f
      Code ID:  12ab2954afbf689002141c2245aa4c3f0f08443e
      Arch:     x86_64
  Contained debug information:
> symtab, debug, unwind
  Usable: yes
```

Running 
```bash
gcc -fuse-ld=gold hello.c -g -o gold-linker-hello.o
```
results in the following output.
```bash
sentry-cli difutil check gold-linker-hello.o
Debug Info File Check
  Type: elf library
  Contained debug identifiers:
> Debug ID: 8fea6672-c8c9-461a-515b-66e5a735a6e4
      Code ID:  7266ea8fc9c81a46515b66e5a735a6e405a807dc
      Arch:     x86_64
  Contained debug information:
> symtab, debug
  Usable: yes
```

## Open questions
1. Is the sentry-cli not compatible with executables linked with gold?

**Comment by jan-auer** [comment]
👋 Sentry engineer here. 

This is a shot in the dark, but could you try passing `--eh-frame-hdr` to the linker? Should look something like this:

```bash
gcc -fuse-ld=gold -Xlinker --eh-frame-hdr hello.c -g -o gold-linker-hello.o
```

If this doesn't work, it would be great if you could either inspect the contents of the unwind section in `default-linker-hello.o`, or share it. Maybe, `gold` removes it because it's not required or not useful. sentry-cli doesn't run such extensive checks.

**Comment by themarwhal** [comment]
Hi @jan-auer! Thanks for commenting :)

The `--eh-frame-hdr` flag did not seem to help.
```bash
vagrant@magma-dev-focal:~$ gcc -fuse-ld=gold -Xlinker --eh-frame-hdr hello.c -g -o gold-linker-hello.o
vagrant@magma-dev-focal:~$ sentry-cli difutil check gold-linker-hello.o
Debug Info File Check
  Type: elf library
  Contained debug identifiers:
    > Debug ID: 1c0e3110-c275-5ddf-b579-c32cd69057ea
      Code ID:  10310e1c75c2df5db579c32cd69057ea12a358bf
      Arch:     x86_64
  Contained debug information:
    > symtab, debug
  Usable: yes
```

Header summary of .o file linked with gold
```bash
vagrant@magma-dev-focal:~$ readelf -S  gold-linker-hello.o
There are 37 section headers, starting at offset 0x1f50:

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .interp           PROGBITS         0000000000000238  00000238
       000000000000001c  0000000000000000   A       0     0     1
  [ 2] .note.ABI-tag     NOTE             0000000000000254  00000254
       0000000000000020  0000000000000000   A       0     0     4
  [ 3] .note.gnu.propert NOTE             0000000000000274  00000274
       0000000000000020  0000000000000000   A       0     0     4
  [ 4] .note.gnu.build-i NOTE             0000000000000294  00000294
       0000000000000024  0000000000000000   A       0     0     4
  [ 5] .dynsym           DYNSYM           00000000000002b8  000002b8
       00000000000000a8  0000000000000018   A       6     1     8
  [ 6] .dynstr           STRTAB           0000000000000360  00000360
       0000000000000082  0000000000000000   A       0     0     1
  [ 7] .gnu.hash         GNU_HASH         00000000000003e8  000003e8
       000000000000001c  0000000000000000   A       5     0     8
  [ 8] .gnu.version      VERSYM           0000000000000404  00000404
       000000000000000e  0000000000000002   A       5     0     2
  [ 9] .gnu.version_r    VERNEED          0000000000000414  00000414
       0000000000000020  0000000000000000   A       6     1     4
  [10] .rela.dyn         RELA             0000000000000438  00000438
       00000000000000c0  0000000000000018   A       5     0     8
  [11] .rela.plt         RELA             00000000000004f8  000004f8
       0000000000000030  0000000000000018  AI       5    13     8
  [12] .init             PROGBITS         0000000000000528  00000528
       000000000000001b  0000000000000000  AX       0     0     4
  [13] .plt              PROGBITS         0000000000000550  00000550
       0000000000000050  0000000000000010  AX       0     0     16
  [14] .text             PROGBITS         00000000000005a0  000005a0
       0000000000000185  0000000000000000  AX       0     0     16
  [15] .fini             PROGBITS         0000000000000728  00000728
       000000000000000d  0000000000000000  AX       0     0     4
  [16] .rodata           PROGBITS         0000000000000738  00000738
       0000000000000012  0000000000000000   A       0     0     4
  [17] .eh_frame         X86_64_UNWIND    0000000000000750  00000750
       00000000000000dc  0000000000000000   A       0     0     8
  [18] .eh_frame_hdr     X86_64_UNWIND    000000000000082c  0000082c
       0000000000000034  0000000000000000   A       0     0     4
  [19] .fini_array       FINI_ARRAY       0000000000001da0  00000da0
       0000000000000008  0000000000000008  WA       0     0     8
  [20] .init_array       INIT_ARRAY       0000000000001da8  00000da8
       0000000000000008  0000000000000008  WA       0     0     8
  [21] .dynamic          DYNAMIC          0000000000001db0  00000db0
       0000000000000200  0000000000000010  WA       6     0     8
  [22] .got              PROGBITS         0000000000001fb0  00000fb0
       0000000000000028  0000000000000000  WA       0     0     8
  [23] .got.plt          PROGBITS         0000000000001fd8  00000fd8
       0000000000000028  0000000000000000  WA       0     0     8
  [24] .data             PROGBITS         0000000000002000  00001000
       0000000000000010  0000000000000000  WA       0     0     8
  [25] .tm_clone_table   PROGBITS         0000000000002010  00001010
       0000000000000000  0000000000000000  WA       0     0     8
  [26] .bss              NOBITS           0000000000002010  00001010
       0000000000000001  0000000000000000  WA       0     0     1
  [27] .comment          PROGBITS         0000000000000000  00001010
       000000000000002b  0000000000000001  MS       0     0     1
  [28] .debug_info       PROGBITS         0000000000000000  0000103b
       0000000000000301  0000000000000000           0     0     1
  [29] .debug_abbrev     PROGBITS         0000000000000000  0000133c
       00000000000000cc  0000000000000000           0     0     1
  [30] .debug_aranges    PROGBITS         0000000000000000  00001408
       0000000000000030  0000000000000000           0     0     1
  [31] .debug_line       PROGBITS         0000000000000000  00001438
       0000000000000113  0000000000000000           0     0     1
  [32] .debug_str        PROGBITS         0000000000000000  0000154b
       00000000000002dc  0000000000000001  MS       0     0     1
  [33] .note.gnu.gold-ve NOTE             0000000000000000  00001828
       000000000000001c  0000000000000000           0     0     4
  [34] .symtab           SYMTAB           0000000000000000  00001848
       0000000000000378  0000000000000018          35    21     8
  [35] .strtab           STRTAB           0000000000000000  00001bc0
       000000000000020c  0000000000000000           0     0     1
  [36] .shstrtab         STRTAB           0000000000000000  00001dcc
       000000000000017d  0000000000000000           0     0     1
```

In comparison the readelf output of an executable linked with plain ld looks like this
```
vagrant@magma-dev-focal:~$ readelf -S  default-linker-hello.o
There are 36 section headers, starting at offset 0x41d0:

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .interp           PROGBITS         0000000000000318  00000318
       000000000000001c  0000000000000000   A       0     0     1
  [ 2] .note.gnu.propert NOTE             0000000000000338  00000338
       0000000000000020  0000000000000000   A       0     0     8
  [ 3] .note.gnu.build-i NOTE             0000000000000358  00000358
       0000000000000024  0000000000000000   A       0     0     4
  [ 4] .note.ABI-tag     NOTE             000000000000037c  0000037c
       0000000000000020  0000000000000000   A       0     0     4
  [ 5] .gnu.hash         GNU_HASH         00000000000003a0  000003a0
       0000000000000024  0000000000000000   A       6     0     8
  [ 6] .dynsym           DYNSYM           00000000000003c8  000003c8
       00000000000000a8  0000000000000018   A       7     1     8
  [ 7] .dynstr           STRTAB           0000000000000470  00000470
       0000000000000082  0000000000000000   A       0     0     1
  [ 8] .gnu.version      VERSYM           00000000000004f2  000004f2
       000000000000000e  0000000000000002   A       6     0     2
  [ 9] .gnu.version_r    VERNEED          0000000000000500  00000500
       0000000000000020  0000000000000000   A       7     1     8
  [10] .rela.dyn         RELA             0000000000000520  00000520
       00000000000000c0  0000000000000018   A       6     0     8
  [11] .rela.plt         RELA             00000000000005e0  000005e0
       0000000000000018  0000000000000018  AI       6    24     8
  [12] .init             PROGBITS         0000000000001000  00001000
       000000000000001b  0000000000000000  AX       0     0     4
  [13] .plt              PROGBITS         0000000000001020  00001020
       0000000000000020  0000000000000010  AX       0     0     16
  [14] .plt.got          PROGBITS         0000000000001040  00001040
       0000000000000010  0000000000000010  AX       0     0     16
  [15] .plt.sec          PROGBITS         0000000000001050  00001050
       0000000000000010  0000000000000010  AX       0     0     16
  [16] .text             PROGBITS         0000000000001060  00001060
       0000000000000185  0000000000000000  AX       0     0     16
  [17] .fini             PROGBITS         00000000000011e8  000011e8
       000000000000000d  0000000000000000  AX       0     0     4
  [18] .rodata           PROGBITS         0000000000002000  00002000
       0000000000000012  0000000000000000   A       0     0     4
  [19] .eh_frame_hdr     PROGBITS         0000000000002014  00002014
       0000000000000044  0000000000000000   A       0     0     4
  [20] .eh_frame         PROGBITS         0000000000002058  00002058
       0000000000000108  0000000000000000   A       0     0     8
  [21] .init_array       INIT_ARRAY       0000000000003db8  00002db8
       0000000000000008  0000000000000008  WA       0     0     8
  [22] .fini_array       FINI_ARRAY       0000000000003dc0  00002dc0
       0000000000000008  0000000000000008  WA       0     0     8
  [23] .dynamic          DYNAMIC          0000000000003dc8  00002dc8
       00000000000001f0  0000000000000010  WA       7     0     8
  [24] .got              PROGBITS         0000000000003fb8  00002fb8
       0000000000000048  0000000000000008  WA       0     0     8
  [25] .data             PROGBITS         0000000000004000  00003000
       0000000000000010  0000000000000000  WA       0     0     8
  [26] .bss              NOBITS           0000000000004010  00003010
       0000000000000008  0000000000000000  WA       0     0     1
  [27] .comment          PROGBITS         0000000000000000  00003010
       000000000000002a  0000000000000001  MS       0     0     1
  [28] .debug_aranges    PROGBITS         0000000000000000  0000303a
       0000000000000030  0000000000000000           0     0     1
  [29] .debug_info       PROGBITS         0000000000000000  0000306a
       0000000000000301  0000000000000000           0     0     1
  [30] .debug_abbrev     PROGBITS         0000000000000000  0000336b
       00000000000000cc  0000000000000000           0     0     1
  [31] .debug_line       PROGBITS         0000000000000000  00003437
       0000000000000113  0000000000000000           0     0     1
  [32] .debug_str        PROGBITS         0000000000000000  0000354a
       000000000000028f  0000000000000001  MS       0     0     1
  [33] .symtab           SYMTAB           0000000000000000  000037e0
       0000000000000690  0000000000000018          34    51     8
  [34] .strtab           STRTAB           0000000000000000  00003e70
       0000000000000203  0000000000000000           0     0     1
  [35] .shstrtab         STRTAB           0000000000000000  00004073
       000000000000015a  0000000000000000           0     0     1
```

**Comment by themarwhal** [comment]
@jan-auer I was not sure what other information would be useful to you, so I've attached the following below:
1. Dockerfile and docker-compose files with all the tools installed so that we can be consistent with each other in terms of versions and what not
2. The .o files produced for both ld and gold linkers from my previous experiment


[sentry-linking-unwind.zip](https://github.com/magma/magma/files/8098116/sentry-linking-unwind.zip)

**Comment by jan-auer** [comment]
Ah, the culprit here is that `gold` creates the `.eh_frame` section as type `X86_64_UNWIND` rather than `SHT_PROGBITS`. Sentry and sentry-cli only check for the latter.

I haven't checked yet what the implications of the different section type are.

**Comment by themarwhal** [comment]
> Ah, the culprit here is that `gold` creates the `.eh_frame` section as type `X86_64_UNWIND` rather than `SHT_PROGBITS`. Sentry and sentry-cli only check for the latter.
> 
> I haven't checked yet what the implications of the different section type are.

@jan-auer Thanks for the update! Can I expect a fix in the near future? Does the linked `getsentry/symbolic` change affect the logic that decides whether symbols exist or not on the platform side as well?

**Comment by jan-auer** [comment]
> Does the linked `getsentry/symbolic` change affect the logic that decides whether symbols exist or not on the platform side as well?

Yes, `symbolic` runs inside both sentry-cli and [symbolicator](https://github.com/getsentry/symbolicator) in Sentry. Once we have the fix merged and released, we will push patch releases for sentry-cli. Should be out this week, although it will take until March to arrive in Sentry self-hosted.

**Comment by electronjoe** [comment]
Thanks again for the assistance @jan-auer!

**Comment by jan-auer** [comment]
The fix is released in [sentry-cli 1.73.0](https://github.com/getsentry/sentry-cli/releases/tag/1.73.0) and deployed to Sentry SaaS. It will become available to Sentry self-hosted with version 22.3.0.

**Comment by themarwhal** [comment]
> The fix is released in [sentry-cli 1.73.0](https://github.com/getsentry/sentry-cli/releases/tag/1.73.0) and deployed to Sentry SaaS. It will become available to Sentry self-hosted with version 22.3.0.

Thank you so much @jan-auer ! I will test it out tomorrow morning. 😊

**Comment by themarwhal** [comment]
Confirmed that the new sentry-cli script is able to find the unwind information!
```bash
@themarwhal ➜ /workspaces/magma (master) $ bazel build lte/gateway/c/session_manager:sessiond --config=production
@themarwhal ➜ /workspaces/magma (master) $ sentry-cli difutil check bazel-bin/lte/gateway/c/session_manager/sessiond
Debug Info File Check
  Type: elf library
  Contained debug identifiers:
    > Debug ID: 0fe996de-5d7a-c11e-7aff-8d28798730ea
      Code ID:  de96e90f7a5d1ec17aff8d28798730ea9a3a6843
      Arch:     x86_64
  Contained debug information:
    > symtab, debug, unwind
  Usable: yes
```
